spr/spr.go

```diff
@@ -0,0 +1,326 @@
+func (spr *StakingPriceRecord) LogFieldsShort() log.Fields {
+	return log.Fields{
+		"spr_hash": hex.EncodeToString(spr.SPRHash),
+	}
+}
+
+// SetPegValues assigns currency polling values to the SPR
+func (spr *StakingPriceRecord) SetPegValues(assets polling.PegAssets) {
+	for asset, v := range assets {
+		spr.Assets.SetValue(asset, v.Value)
+	}
+}
+
+// NewSpr collects all the information unique to this staker and its configuration, and also
+// goes and gets the oracle data.
+func NewSpr(ctx context.Context, dbht int32, c *config.Config) (spr *StakingPriceRecord, err error) {
+	spr = NewStakingPriceRecord()
+
+	/**
+	 *	Init SPR with configuration settings
+	 */
+	protocol, err1 := c.String("Staker.Protocol")
+	network, err2 := common.LoadConfigStakerNetwork(c)
+	spr.Network = network
+	spr.Protocol = protocol
+
+	if err1 != nil {
+		return nil, errors.New("config file has no Staker.Protocol specified")
+	}
+	if err2 != nil {
+		return nil, errors.New("config file has no Staker.Network specified")
+	}
+	spr.SPRChainID = base58.Encode(common.ComputeChainIDFromStrings([]string{protocol, network, common.SPRChainTag}))
+	spr.Dbht = dbht
+	spr.Version = common.SPRVersion(spr.Network, int64(spr.Dbht))
+
+	if network == common.TestNetwork {
+		fct := common.DebugFCTaddresses[0][1]
+		spr.CoinbaseAddress = fct
+	} else {
+		if str, err := c.String("Staker.CoinbaseAddress"); err != nil {
+			return nil, errors.New("config file has no Coinbase PEG Address")
+		} else {
+			spr.CoinbaseAddress = str
+		}
+	}
+
+	spr.CoinbasePEGAddress, err = common.ConvertFCTtoPegNetAsset(network, "PEG", spr.CoinbaseAddress)
+	if err != nil {
+		log.Errorf("invalid fct address in config file: %v", err)
+	}
+
+	/**
+	 *	Get SPR Record with Assets data (polling)
+	 */
+	err = spr.GetSPRecord(c)
+	if err != nil {
+		return nil, err
+	}
+
+	if !spr.Validate(c, int64(dbht)) {
+		if !common.NetworkActive(spr.Network, int64(dbht)) {
+			return nil, fmt.Errorf("Waiting for activation height")
+		}
+		return nil, fmt.Errorf("spr invalid")
+	}
+	return spr, nil
+}
+
+// GetSPRecord initializes the SPR with polling data and factom entry
+func (spr *StakingPriceRecord) GetSPRecord(c *config.Config) error {
+	InitDataSource(c) // Kinda odd to have this here.
+
+	oprVersion := 1
+	if spr.Version == 0 {
+		oprVersion = 4
+	}
+
+	//get asset values
+	Peg, err := PollingDataSource.PullAllPEGAssets(uint8(oprVersion))
+	if err != nil {
+		return err
+	}
+	spr.SetPegValues(Peg)
+
+	data, err := spr.SafeMarshal()
+	if err != nil {
+		panic(err)
+	}
+	sha := sha256.Sum256(data)
+	spr.SPRHash = sha[:]
+	return nil
+}
+
+// CreateSPREntry will create the entry from the EXISITING data.
+// It will not set any fields like in `GetSPRecord`
+func (spr *StakingPriceRecord) CreateSPREntry() (*factom.Entry, error) {
+	var err error
+	e := new(factom.Entry)
+	e.ChainID = hex.EncodeToString(base58.Decode(spr.SPRChainID))
+
+	//  Todo: verify every options
```

Comment:
Is there an update to this? Where is option verification code?

Reply:
Do you see this PR is Draft? It is for verify options for SPREntry, @sigrlami

Reply:
I have new interface and didn't noticed it's a draft. Thanks for pointing out.